### ISSUE TITLE
clang issues on ubuntu 24.04 and Python 3.12

### ISSUE THREAD

**Issue by tunguz** [exo-explore/exo#458]
![Screenshot from 2024-11-14 08-18-52](https://github.com/user-attachments/assets/c292fece-2c1e-4d5b-b3f2-0c82219ed096)

After installing exo and clang on an Ubuntu 24.04 machine with a Ryzen CPU I got an error while trying to run a prompt. (See attached image.) Anyone have any idea what might be going on?

**Comment by devinatkin** [comment]
I got this same issue myself. Currently trying my hand at diagnosing what might be the cause. I'm guess it's something to do with the fact that it's running in a venv with python there. 

![image](https://github.com/user-attachments/assets/82136d24-f683-475f-ae59-08e8685caa49)

I'm on a beefy setup so I know it's not a resource constraint issue (2 1080 Ti)

**Comment by tunguz** [comment]
@devinatkin I don't think its the venv issue. I am running it on bare metal Ubuntu 24.04, which comes with Python 3.12 as the default system Python.

**Comment by devinatkin** [comment]
> @devinatkin I don't think its the venv issue. I am running it on bare metal Ubuntu 24.04, which comes with Python 3.12 as the default system Python.

Well that's good to know I'm on a fresh install of Ubuntu 24.04.1 LTS and decided to try with just the 1 machine pretty good machine before adding the rest of the junk heap.

**Comment by tunguz** [comment]
> > @devinatkin I don't think its the venv issue. I am running it on bare metal Ubuntu 24.04, which comes with Python 3.12 as the default system Python.
> 
> Well that's good to know I'm on a fresh install of Ubuntu 24.04.1 LTS and decided to try with just the 1 machine pretty good machine before adding the rest of the junk heap.

Yup, I am using a brand new machine with a brand new Ubuntu install. This was pretty much the first thing I had tried on it.

**Comment by cadenmackenzie** [comment]
+1 I was getting this as well. Thought maybe the clang version wasn't compatible with the current tinygrad implementation so tried clang 14 and 16 but couldn't get it to work.

**Comment by tunguz** [comment]
> Basically some kind of path issue as it has a repeating component, that shouldn't be repeating. it is compiling from pipe, not a file, so I can't debug further

Would it be possible to overcome this by a different kind of clang installation?

**Comment by jonstelly** [comment]
I'm having other clang problems trying to get a docker image together ([Dockerfile](https://github.com/smahtz/exo/blob/feature/docker/Dockerfile) if curious) so this might just change from the issue you're seeing to the one I'm seeing, but to the question of trying other clang installation methods...

https://apt.llvm.org/ for nightly builds, other versions, etc...

would be curious to hear if you get things working with a different version / install.

**Comment by AlexCheema** [comment]
@blindcrone you're running this on your linux box right? Could you take a look at what might be the issue here? Thanks!

**Comment by blindcrone** [comment]
I've only got the tinygrad backend working on linux machines that have GPUs. I chased this rabbithole a bit and found that this is an issue in tinygrad in general, as I've yet to find any report of a linux user being able to use the LLVM
backend on tinygrad for llama and related models. Tinygrad contains a "fix-bf16"-alike function that also doesn't seem to solve the issue.

The actual bug is happening in LLVM, and is an issue I was able to chase down in that repository, I'll look for it again and find links to post here, but the tl;dr is that this might be patched in LLVM 19 but no distro I know of currently packages it because of build issues

**Comment by lexasub** [comment]
@blindcrone may be try on archlinux? https://archlinux.pkgs.org/rolling/mesa-git-x86_64/clang-git-20.0.0_r516907.e102338b6e2f-1-x86_64.pkg.tar.zst.html

**Comment by lexasub** [comment]
@blindcrone  i install it(clang 20), but is not fixed

**Comment by lexasub** [comment]
@blindcrone after clean ccache cache, i another problem, exo didn't use gpu (cluster mode - 2 linux machines, rtx 3060). and i don't get answer in ui
[log.log](https://github.com/user-attachments/files/17841846/log.log)

**Comment by kdkd** [comment]
I'm attempting this on aarch64(raspberry pi 5). With stock clang 14, it's failing with `error: __bf16 is not supported on this target`. I upgraded to clang 18, which is blowing up in a weirder place:

```
fatal error: error in backend: Cannot select: 0x555591ddaa60: f16 = fp_round 0x555591ddac90, TargetConstant:i64<0>
  0x555591ddac90: bf16,ch = load<(load (s16) from %ir.7 + 4, !tbaa !4)> 0x555591d763e0, 0x555591dd63e0, undef:i64
    0x555591dd63e0: i64 = add nuw 0x555591dd5960, Constant:i64<4>
      0x555591dd5960: i64 = add 0x555591dd59d0, 0x555591dd5a40
        0x555591dd59d0: i64,ch = CopyFromReg 0x555591d763e0, Register:i64 %3
          0x555591dd5ab0: i64 = Register %3
        0x555591dd5a40: i64 = shl nuw nsw 0x555591dd57a0, Constant:i64<3>
          0x555591dd57a0: i64,ch = CopyFromReg 0x555591d763e0, Register:i64 %0
            0x555591dd5810: i64 = Register %0
          0x555591dd5b20: i64 = Constant<3>
      0x555591dd61b0: i64 = Constant<4>
    0x555591dd58f0: i64 = undef
  0x555591dd5c00: i64 = TargetConstant<0>
In function: E_4194304_4
clang-18: error: clang frontend command failed with exit code 70 (use -v to see invocation)
Debian clang version 18.1.8 (++20240731024826+3b5b5c1ec4a3-1~exp1~20240731144843.145)
```

With clang 19, it builds and executes okay, but I still don't have it working because i'm getting socket errors immediately after that that I haven't debugged yet.

**Comment by Coastline-3102** [comment]
I believe I am also experiencing this error. I am on Debian 12 (bookworm) with Python 3.12.7 and clang version 14.0.6

Is there a "known good" combination of distro and python/clang versions that work? I have been testing out my own version of a Dockerfile so I can deploy this to multiple systems, but that is also getting the same error.

**Comment by DrEVILish** [comment]
Tested using an LXC (proxmox cluster) installed Debian 12, Python 3.11.2, Clang 14.0.6

```
root@exo-1:~# clang -v
Debian clang version 14.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/12
Selected GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/12
Candidate multilib: .;@m64
Selected multilib: .;@m64
```
```
root@exo-1:~# python3
Python 3.11.2 (main, Sep 14 2024, 03:00:30) [GCC 12.2.0] on linux
```
```
root@exo-1:~# uname -a
Linux exo-1 6.8.12-2-pve #1 SMP PREEMPT_DYNAMIC PMX 6.8.12-2 (2024-09-05T10:03Z) x86_64 GNU/Linux
```

**Comment by lexasub** [comment]
I just did a git pull, and it seems like Exo started working in the cluster on two machines after installation

**Comment by tunguz** [comment]
@lexasub Were those Linux CPU-only machines? What is your setup?

**Comment by lexasub** [comment]
@tunguz, no rtx 3060 on two machines. @blindcrone, why on cluster mode available only llama?

**Comment by lexasub** [comment]
clang --version
clang version 19.1.0 (Fedora 19.1.0-1.fc41)

**Comment by tunguz** [comment]
@lexasub Yeah, exo seems to work fine on GPU Linux computers. I’d like to see it enabled on the CPU-only machines as well. That’s what this issue is all about. :/

**Comment by lexasub** [comment]
@tunguz The first time I wrote here, I encountered the same problem as you.

**Comment by Coastline-3102** [comment]
> Yeah, exo seems to work fine on GPU Linux computers

I suspect that might be just NVIDA GPU Linux computers that it works fine on. The computer I have been testing on has an AMD GPU.

@lexasub if I am reading your comments right, you are running two separate 3060 computers, had this issue, but were able to fix it? What did you do to fix the issue?

**Comment by lexasub** [comment]
@Coastline-3102, yes, you right. but llama3.1 use small resources on gpu, than peek computation, may be it not solved(, but in htop i didn't see big cpu usage, may be computation was been on gpu

**Comment by DrEVILish** [comment]
> @DrEVILish, you mentioned having it working on LXC/proxmox cluster. What CPU/GPU does your system have?

Sorry, tested, and not working, I've updated my comment above. It was a container without a "GPU" attached just a CPU instance.

**Comment by terasakisatoshi** [comment]
This `Dockerfile` as below reproduced errors reported here.

https://github.com/AtelierArith/DocstringTranslationExoBackend.jl/blob/main/docker/Dockerfile

Just run the following command

```sh
git clone https://github.com/AtelierArith/DocstringTranslationExoBackend.jl.git
cd DocstringTranslationExoBackend.jl
cd docker
docker build -t exojl -f Dockerfile . &&  docker run --rm -it -p 52415:52415 exojl
```

Then open your browser and navigate to localhost:52415

`Say something`.

**Comment by crimson-hawk5** [comment]
Its not working on CPU, Raspberry pi 5's. Here is the error:
"WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1733215714.711032   67196 fork_posix.cc:75] Other threads are currently calling into gRPC, skipping fork() handlers
clang: error: the clang compiler does not support '-march=native'"

**Comment by benson-chiu** [comment]
Most of this question points to the fact that not all CPUs support the `bfloat16` format.
As @blindcrone said, this is indeed a problem in tinygrad, where I found this pull request:
https://github.com/tinygrad/tinygrad/pull/7032/files

I try to replace the `fix_bf16` function in `exo/inference/tinygrad/models/llama.py` with this code:
``` python
def fix_bf16(weights:Dict[Any, Tensor]):
  if Device.DEFAULT == "CLANG":
    # TODO: without casting to float16, 70B llama OOM on tinybox.
    return {
      k: (v.llvm_bf16_cast(dtypes.float32).to(v.device) if v.dtype == dtypes.bfloat16 else v) 
      for k, v in weights.items()
    }
  # TODO: check if device supports bf16
  return {k:v.cast(dtypes.float16) if v.dtype == dtypes.bfloat16 else v for k,v in weights.items()}
```

Only using CPU to execute exo, it works successfully.
![image](https://github.com/user-attachments/assets/b040c1b5-2281-4cb0-9b21-aab9a6c8a242)

**Comment by DrEVILish** [comment]
Changing this fixed the CLANG issue for me.

Then got a reported issue about `No module named 'llvmlite'` solved by `pip install numba`

Suggest creating a PR for the above changes.

**Comment by macauleycheng** [comment]
pip3 install llvmlite

**Comment by tunguz** [comment]
![Screenshot from 2024-12-05 06-48-04](https://github.com/user-attachments/assets/90ca05bb-eb09-4228-91a9-b52e0d82a8d5)
Yay, it's working!

**Comment by blindcrone** [comment]
Has anyone gotten more than one prompt to go through this way? I did these fixes but it took quite a while to get one prompt through, then hung indefinitely on any subsequent prompt. I'd chalk the first thing up to CPU simply being slow for inference, but the second thing seems like a deeper issue

**Comment by wbic16** [comment]
I can confirm that the tips on this issue work on a stock AMD system to get exo into a working state.

I had to install these packages in addition to patching tinygrad:
https://github.com/wbic16/exocortical/blob/exo/setup.sh

![image](https://github.com/user-attachments/assets/4477ea76-758d-47e7-96c6-2433974b0337)

**Comment by wbic16** [comment]
I opened a PR for the CPU/clang issue. Thanks @benson-chiu for the tip!
https://github.com/exo-explore/exo/pull/542

**Comment by wbic16** [comment]
> Has anyone gotten more than one prompt to go through this way? I did these fixes but it took quite a while to get one prompt through, then hung indefinitely on any subsequent prompt. I'd chalk the first thing up to CPU simply being slow for inference, but the second thing seems like a deeper issue

I posted a screenshot of multiple replies after applying that tinygrad patch. I'm running a single AMD 8945HS with a Radeon 780M and 96 GB of RAM in that example. It takes a while, but does seem to be working.

**Comment by terasakisatoshi** [comment]
> Its not working on CPU, Raspberry pi 5's. Here is the error: "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR I0000 00:00:1733215714.711032 67196 fork_posix.cc:75] Other threads are currently calling into gRPC, skipping fork() handlers clang: error: the clang compiler does not support '-march=native'"

I still have the same issue on my 4 x RaspberryPi5 even though I installed the numba and llvmlite

<img width="1077" alt="image" src="https://github.com/user-attachments/assets/441a13b0-1633-4b4f-a6c7-860e72ec0645" />
